### PR TITLE
[SPARK-23218][SQL] simplify ColumnVector.getArray

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

`ColumnVector` is very flexible about how to implement array type. As a result `ColumnVector` has 3 abstract methods for array type: `arrayData`, `getArrayOffset`, `getArrayLength`. For example, in `WritableColumnVector` we use the first child vector as the array data vector, and store offsets and lengths in 2 arrays in the parent vector. `ArrowColumnVector` has a different implementation.

This PR simplifies `ColumnVector` by using only one abstract method for array type: `getArray`.

## How was this patch tested?

existing tests.

rerun `ColumnarBatchBenchmark`, there is no performance regression.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @hvanhovell @ueshin @viirya @gatorsmile

**Comment by ueshin** [comment]
LGTM pending Jenkins.

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java [hunk 1]
```diff
@@ -450,13 +439,11 @@ final boolean isNullAt(int rowId) {
...
-      return accessor.getInnerValueCountAt(rowId);
-    }
-
-    @Override
-    final int getArrayOffset(int rowId) {
-      return accessor.getOffsetBuffer().getInt(rowId * accessor.OFFSET_WIDTH);
+    final ColumnarArray getArray(int rowId) {
+      int index = rowId * accessor.OFFSET_WIDTH;
+      int start = offsets.getInt(index);
+      int end = offsets.getInt(index + accessor.OFFSET_WIDTH);
```
Comment (kiszk) [inline sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java]: Does this code work when `rowId == arrayLength`? In other words, is the length of `offsetBuffer` equal to `(rowId + 1) * accessor.OFFSET_WIDTH`?
cc: @ueshin

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchBenchmark.scala [hunk 2]
```diff
@@ -479,10 +475,10 @@ object ColumnarBatchBenchmark {
...
     Array Vector Read:                       Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
     ------------------------------------------------------------------------------------------------
-    On Heap Read Size Only                         416 /  423        393.5           2.5       1.0X
-    Off Heap Read Size Only                        396 /  404        413.6           2.4       1.1X
-    On Heap Read Elements                         2569 / 2590         63.8          15.7       0.2X
-    Off Heap Read Elements                        3302 / 3333         49.6          20.2       0.1X
+    On Heap Read Size Only                         426 /  437        384.9           2.6       1.0X
+    Off Heap Read Size Only                        406 /  421        404.0           2.5       1.0X
+    On Heap Read Elements                         2636 / 2642         62.2          16.1       0.2X
+    Off Heap Read Elements                        3770 / 3774         43.5          23.0       0.1X
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchBenchmark.scala]: Thank you for updating this in the PR. It seems to cause a little performance degradation.
Could you update `OrcReadBenchmark` and `FilterPushdownBenchmark`, too?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchBenchmark.scala]: orc and parquet vectorized reader don't support array type yet.

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java]: I believe it is, by looking at `accessor.getInnerValueCountAt`. cc @BryanCutler for confirmation.

**Comment by cloud-fan** [comment]
retest this please

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java [hunk 3]
```diff
@@ -433,10 +418,14 @@ final long getLong(int rowId) {
   private static class ArrayAccessor extends ArrowVectorAccessor {
 
     private final ListVector accessor;
+    private final ArrowBuf offsets;
+    private final ArrowColumnVector arrayData;
 
     ArrayAccessor(ListVector vector) {
       super(vector);
       this.accessor = vector;
+      this.offsets = vector.getOffsetBuffer();
```
Comment (ueshin) [inline sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java]: Maybe `vector.getOffsetBuffer()` will change when loading the next batch so we can't reuse the `ArrowBuf`.

Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java]: good catch!

**Comment by kiszk** [comment]
retest this please

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java [hunk 4]
```diff
@@ -182,57 +187,54 @@
...
   public final ColumnarRow getStruct(int rowId) {
     return new ColumnarRow(this, rowId);
   }
 
   /**
-   * Returns the array for rowId.
+   * Returns the array type value for rowId.
+   *
+   * To support array type, implementations must construct an {@link ColumnarArray} and return it in
+   * this method. {@link ColumnarArray} requires a {@link ColumnVector} that stores the data of all
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java]: Shall we move the details of `ColumnarArray` implementation to `ColumnarArray` class?

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java [hunk 5]
```diff
@@ -450,13 +437,12 @@ final boolean isNullAt(int rowId) {
...
-    }
-
-    @Override
-    final int getArrayOffset(int rowId) {
-      return accessor.getOffsetBuffer().getInt(rowId * accessor.OFFSET_WIDTH);
+    final ColumnarArray getArray(int rowId) {
+      ArrowBuf offsets = accessor.getOffsetBuffer();
+      int index = rowId * accessor.OFFSET_WIDTH;
+      int start = offsets.getInt(index);
+      int end = offsets.getInt(index + accessor.OFFSET_WIDTH);
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java]: At the last rowId, what `offsets.getInt(index + accessor.OFFSET_WIDTH)` returns?

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java [hunk 6]
```diff
@@ -602,7 +603,17 @@ public final int appendStruct(boolean isNull) {
...
+    return new ColumnarArray(arrayData(), getArrayOffset(rowId), getArrayLength(rowId));
+  }
+
+  public WritableColumnVector arrayData() {
+    return childColumns[0];
+  }
+
+  public abstract int getArrayLength(int rowId);
+
+  public abstract int getArrayOffset(int rowId);
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java]: Shall we make these two methods private/protected?

**Comment by viirya** [comment]
LGTM with few minor comments.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java (hunk 6 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java]: `WritableColumnVector` is an internal class, so I want to keep the visibility wider for the ease of use.

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java (hunk 5 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java]: the end offset of the last record. see https://arrow.apache.org/docs/memory_layout.html#list-type

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/vectorized/ColumnVector.java]: The comments here are mostly about how to construct the `ColumnarArray`, but about `ColumnarArray` itself, so I'd like to keep them here.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java (hunk 6 above)
Comment (viirya) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/WritableColumnVector.java]: Ok. Make sense.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchBenchmark.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/vectorized/ColumnarBatchBenchmark.scala]: Right. I see.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by gatorsmile** [comment]
Thanks! Merged to master/2.3

File: sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java (hunk 1 above)
Comment (BryanCutler) [reply sql/core/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java]: yes, the code is correct.  `offsetBuffer` will be sized to at least `(numRows + 1) * accessor.OFFSET_WIDTH`

**Pull request closed without merging**